### PR TITLE
[Merged by Bors] - UCB: hare generate unified content block

### LINKED ISSUE #2903: Create `ProposalStore` under Hare
Today, when a block is processed, it's stored in the database. Every time the clock ticks, the Hare waits `WakeupDelta` seconds and then calls `mesh.LayerBlocks(lyrID)` to get the list of on-time blocks from the database.

This mechanism should be replaced by a `ProposalStore` managed by the Hare.

The `ProposalStore` should support:
- Storing a `Proposal`
- Getting a list of `Proposal`s for a given layer
- Fetching a specific `Proposal` (given `LayerID` and `ProposalID`)
- Pruning all `Proposal`s from before a specific layer

The `ProposalStore` should be a LevelDB. The key should be the concatenation of `LayerID+ProposalID`. Layers more than `hdist` back should be pruned regularly.

Once the `ProposalStore` is implemented we should replace any usage of the `BallotDB` to the `ProposalStore` if the proposal (list of transactions) is actually needed. The Hare should agree on a set of `ProposalID`s and executing layers will be done with proposals, but the Tortoise should use ballots.

## Why not keep it in memory?
I originally thought of just keeping it in memory as `map[LayerID]map[ProposalID]Proposal`. There should be `blocks_per_layer * hdist` proposals in the store. Not sure about the final values, but I think reasonable numbers for scale are `50 * 120 = 6000`. The size of a `Proposal` will mainly depend on the number of transactions that we allow referencing in one (assuming we strip the embedded `Ballot`), each `TransactionID` is currently 32-bytes, but we may decide to reduce it to 20-bytes. If we allow 100 transactions per proposal, this comes out to <20mb which is not negligible to keep in memory, but still viable.

I'm assuming 20% "wasted" transactions (duplicates, eventually invalid, etc.) which means `100 transactions per proposal * 50 blocks per layer * 80% = 4000` then, dividing by 300 seconds per layer brings us to ~13 transactions per second. I think we're aiming for higher TPS, so we should probably store the proposals on disk.

### LINKED ISSUE #2904: Ability to Fetch `Proposal`s from Peers
Eventually we'll want all existing mechanisms for blocks to work for `Ballot`s. However, in addition to the ability to sync ballots (which is already implemented for blocks today) we also need to the ability to fetch a specific `Proposal` from peers if we see it referenced in a Hare message and we don't have it.

There should never be a need to get a list of proposals. We only need the ability to fetch a specific proposal from peers, given `LayerID` and `ProposalID`. Proposals will be fetched from the `ProposalStore`.

The fetching mechanism will be used when implementing `Block` derivation.

### LINKED ISSUE #2905: Store Only Ballots in `BallotDB` and Use `BallotID` for Referring to Ballots
At this point we can start saving only the `Ballot` that's embedded in the `Proposal` in the `BallotDB`, index the database using `BallotID`s and change references to ballots so they use `BallotID`s.

### LINKED ISSUE #2906: Derive `Block` from Hare Output Set
When the Hare completes we'll now want to generate a `Block` and store it in a separate database.

We'll use the same mechanism we use today to create a list of ordered transactions (`mesh.extractUniqueOrderedTransactions()`), to generate a `Block`. In addition to an ordered list of transactions, a block should contain a list of rewards. This should be derived from the list of smeshers that authored the proposals that the Hare agreed on. For each smesher we'll find the coinbase address in the relevant ATX and add a reward proportional to the number of proposals/ballots authored by them and included in the Hare output set.

Blocks should be persisted to a `BlockDB`. When the node calculates a block from Hare results, it should also store a value in a `layer -> block` index in the same database. To be explicit - we should NOT store a value in this index if we add a block to the database because we've synced it or fetched it from a peer, only when we've created it ourselves based on Hare results.

### LINKED ISSUE #2907: Update Voting to Blocks
When publishing a `Ballot` we should compare the index of blocks we've derived ourselves with the base block. We should have derived one block for every layer in `hdist` and the lists should match. If they don't we should add appropriate diffs. If we don't have a derived block for a layer within `hdist` - we should vote neutral for it.

Now that votes are for blocks, the state transition function (`mesh.applyState`) should no longer derive the list of transactions itself, and instead it should pick the block with sufficient votes and execute the rewards and transactions from it in order.

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

Closes #2903
Closes #2904
Closes #2905
Closes #2906
Closes #2907

## Changes

- hare: 
  generate unified content block from proposals in the consensus output
- mesh
  when apply a layer, it picks the lowest BlockID among all the valid blocks verified by tortoise
- tortoise
  within hdist, vote Support for hare output block only, and vote Against for any other blocks in the layer
  outside hdist, no change.

## Test Plan

unit tests, system tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

File: tortoise/tortoise.go [hunk 1]
```diff
@@ -761,8 +761,12 @@ func (t *turtle) addLocalVotes(ctx context.Context, logger log.Log, lid types.La
 			t.undecided[lid] = struct{}{}
 			return nil
 		}
-		for _, bid := range hareOutput {
-			t.hareOutput[bid] = support
+		bids := t.blocks[lid]
+		for _, bid := range bids {
```
Comment (dshulyak) [inline tortoise/tortoise.go]: please remove this loop. if hare doesn't support a block it should not be added to the t.hareOutput

File: tortoise/tortoise.go [hunk 2]
```diff
@@ -746,8 +746,8 @@ func (t *turtle) addLocalVotes(ctx context.Context, logger log.Log, lid types.La
 	)
 
 	if !lid.Before(t.layerCutoff()) {
-		// for newer layers, we vote according to the local opinion (hare output, from hare or sync)
-		hareOutput, err := t.bdp.GetHareConsensusOutput(lid)
+		// for newer layers, we vote according to the local opinion (hare output, from live participation or sync)
+		theOne, err := t.bdp.GetHareConsensusOutput(lid)
```
Comment (dshulyak) [inline tortoise/tortoise.go]: hareOutput is better variable

File: tortoise/tortoise.go [hunk 3]
```diff
@@ -761,8 +761,12 @@ func (t *turtle) addLocalVotes(ctx context.Context, logger log.Log, lid types.La
 			t.undecided[lid] = struct{}{}
 			return nil
 		}
-		for _, bid := range hareOutput {
-			t.hareOutput[bid] = support
+		bids := t.blocks[lid]
+		for _, bid := range bids {
+			t.hareOutput[bid] = against
+		}
+		if theOne != types.EmptyBlockID {
```
Comment (dshulyak) [inline tortoise/tortoise.go]: when block will be empty and error is nil?

File: tortoise/sim/layer.go [hunk 4]
```diff
@@ -142,39 +132,44 @@ func (g *Generator) genLayer(cfg nextConf) types.LayerID {
...
+		if err = ballot.Initialize(); err != nil {
+			g.logger.With().Panic("failed to init ballot", log.Err(err))
+		}
+		for _, state := range g.states {
+			state.OnBallot(ballot)
+		}
+		layer.AddBallot(ballot)
+	}
+	for i := 0; i < numBlocks; i++ {
+		block := types.GenLayerBlock(g.nextLayer, types.RandomTXSet(100))
```
Comment (dshulyak) [inline tortoise/sim/layer.go]: please use seeded randomness, this helpers cannot be used

File: tortoise/rerun_test.go [hunk 5]
```diff
@@ -101,10 +102,13 @@ func testWindowCounting(tb testing.TB, maliciousLayers, verifyingWindow, fullWin
 	}
 	require.NoError(tb, tortoise.rerun(ctx))
 
-	for _, block := range blocks {
-		validity, err := s.GetState(0).MeshDB.ContextualValidity(block)
-		require.NoError(tb, err)
-		require.Equal(tb, expectedValidity, validity, "validity for block %s", block)
+	validBlocks, err := s.GetState(0).MeshDB.LayerContextuallyValidBlocks(context.TODO(), misverified)
```
Comment (dshulyak) [inline tortoise/rerun_test.go]: what was the reason for this change? i think it needs to be reverted

this is actually completely incorrect now. all blocks must be verified

**Review by dshulyak — CHANGES_REQUESTED** [review]

File: tortoise/tortoise_test.go [hunk 6]
```diff
@@ -1377,17 +1408,18 @@ func skipLayers(n int) sim.VotesGenerator {
 			panic(fmt.Sprintf("need at least %d layers", position))
 		}
 		baseLayer := layers[len(layers)-position]
-		support := layers[len(layers)-position].BlocksIDs()
+		support := layers[len(layers)-position].BlocksIDs()[0:1]
```
Comment (dshulyak) [inline tortoise/tortoise_test.go]: please revert this change. it doesn't have anything to do with hare

File: tortoise/tortoise_test.go [hunk 7]
```diff
@@ -1698,11 +1730,13 @@ func TestVoteAgainstSupportedByBaseBallot(t *testing.T) {
 
 	unsupported := map[types.BlockID]struct{}{}
 	for lid := genesis; lid.Before(last); lid = lid.Add(1) {
-		bids, err := s.GetState(0).MeshDB.LayerBlockIds(lid)
+		theOne, err := s.GetState(0).MeshDB.GetHareConsensusOutput(lid)
```
Comment (dshulyak) [inline tortoise/tortoise_test.go]: please rename to hareOutput

File: tortoise/tortoise_test.go [hunk 8]
```diff
@@ -1698,11 +1730,13 @@ func TestVoteAgainstSupportedByBaseBallot(t *testing.T) {
...
 		require.NoError(t, err)
-		tortoise.trtl.validity[bids[0]] = against
-		tortoise.trtl.hareOutput[bids[0]] = against
-		unsupported[bids[0]] = struct{}{}
+		if theOne != types.EmptyBlockID {
```
Comment (dshulyak) [inline tortoise/tortoise_test.go]: this conditiona should be removed

File: tortoise/tortoise_test.go [hunk 9]
```diff
@@ -1804,7 +1838,18 @@ func TestComputeLocalOpinion(t *testing.T) {
 			require.NoError(t, err)
 			for _, bid := range blocks {
 				vote, _ := getLocalVote(&tortoise.trtl.commonState, cfg, tc.lid, bid)
-				require.Equal(t, tc.expected, vote, "block id %s", bid)
+				if tc.expected == support {
+					theOne, err := s.GetState(0).MeshDB.GetHareConsensusOutput(tc.lid)
```
Comment (dshulyak) [inline tortoise/tortoise_test.go]: pleased rename to hare output

File: tortoise/rerun_test.go (hunk 5 above)
Comment (dshulyak) [reply tortoise/rerun_test.go]: this are the changes to fix it https://github.com/dshulyak/go-spacemesh/commit/8d7067db9f640eb270ebb987c104e71a039dfb90

File: tortoise/sim/layer.go (hunk 4 above)
Comment (countvonzero) [reply tortoise/sim/layer.go]: i am wary of disturbing the random sequence in sim package.
when i tried using `g.rng` some tests randomly failed. for example, if i use `g.genTXIDs(5)` TestBaseBallotEvictedBlock passes but `g.genTXIDs(100)` it fails. this is just one example.

File: tortoise/tortoise.go (hunk 3 above)
Comment (countvonzero) [reply tortoise/tortoise.go]: when hare output empty set

File: tortoise/rerun_test.go (hunk 5 above)
Comment (countvonzero) [reply tortoise/rerun_test.go]: thanks for the patch. applied

File: tortoise/sim/layer.go (hunk 4 above)
Comment (dshulyak) [reply tortoise/sim/layer.go]: i guess it has something to do with random weights, i thought they were not used by default anymore. nvm then i will take a look after

File: tortoise/tortoise_test.go (hunk 8 above)
Comment (countvonzero) [reply tortoise/tortoise_test.go]: this is due to the explicit `sim.WithEmptyHareOutput()`. so one of the valid block will be types.EmptyBlockID which won't be in the support list.

File: collector/collector.go [hunk 10]
```diff
@@ -113,6 +120,17 @@ func (c *EventsCollector) collectEvents(url string) {
 loop:
 	for {
 		select {
+		case data := <-proposals:
```
Comment (dshulyak) [inline collector/collector.go:123]: what this collector does? maybe this is some old artifact and we can ditch it instead of maintaining?

File: common/types/block.go [hunk 11]
```diff
@@ -1,41 +1,100 @@
...
+type Block struct {
+	InnerBlock
+	// the following fields are kept private and from being serialized
+	blockID BlockID
+}
+
+// InnerBlock contains the transactions and rewards of a block.
+type InnerBlock struct {
+	LayerIndex LayerID
+	Rewards    []AnyReward
```
Comment (dshulyak) [inline common/types/block.go:34]: type Reward is reserved for something else?

File: mesh/mesh.go [hunk 12]
```diff
@@ -406,19 +399,43 @@ func (msh *Mesh) pushLayersToState(ctx context.Context, oldVerified, newVerified
...
+		if len(blocks) > 0 {
+			// when tortoise verify multiple blocks in the same layer, we only apply one with the lowest
+			// lexicographical sort order
+			applied = blocks[0]
+			if len(blocks) > 1 {
+				notApplied = append(notApplied, blocks[1:]...)
+			}
+		}
+
+		if err = msh.updateStateWithLayer(ctx, layerID, applied); err != nil {
```
Comment (dshulyak) [inline mesh/mesh.go:423]: is it expected that applied will be nil is layer is empty?

File: mesh/mesh.go [hunk 13]
```diff
@@ -431,133 +448,128 @@ func (msh *Mesh) revertState(ctx context.Context, layerID types.LayerID) error {
...
-	lyr := types.NewExistingLayer(validatedLayer, blocks)
-	msh.ProcessLayer(ctx, lyr)
+	return msh.ProcessLayer(ctx, layerID)
 }
 
-// apply the state for a single layer. stores layer results for early layers, and applies previously stored results for
-// now-older layers.
-func (msh *Mesh) updateStateWithLayer(ctx context.Context, layer *types.Layer) {
+// apply the state for a single layer.
+func (msh *Mesh) updateStateWithLayer(ctx context.Context, layerID types.LayerID, hareOutput *types.Block) error {
```
Comment (dshulyak) [inline mesh/mesh.go]: i think it doesn't make sense to call it hare output after tortoise verified a block

**Review by dshulyak — APPROVED** [review]

File: mesh/mesh.go (hunk 12 above)
Comment (countvonzero) [reply mesh/mesh.go:423]: yes. if the hare output an empty set, then it's likely that the tortoise will not verify any block. i.e. no blocks in that layer has contextual validity bit set to true.

File: common/types/block.go (hunk 11 above)
Comment (countvonzero) [reply common/types/block.go:34]: yes. it's used in other place. i meant for whoever that works on TX selection/rewards to re-arrange the naming.

File: collector/collector.go (hunk 10 above)
Comment (countvonzero) [reply collector/collector.go:123]: the only thing is the memory collector used in cmd/multi_node_sim.go. i am leaving it for now. not sure what it's for tho.

File: mesh/mesh.go (hunk 13 above)
Comment (countvonzero) [reply mesh/mesh.go]: changed to `block`

**Comment by countvonzero** [comment]
bors merge
